### PR TITLE
[IZPACK-1788] created installer is installed without classifier

### PR DESCRIPTION
**Pull request by hiteshbosamiya**

### CODE REVIEW

File: izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java [hunk 1]
```diff
@@ -181,22 +191,49 @@ public void execute() throws MojoExecutionException, MojoFailureException
...
+            Artifact artifact = project.getArtifact();
+            artifact.setFile(jarFile);
+        }
+        else
+        {
+            if (enableAttachArtifact)
+            {
+                projectHelper.attachArtifact(project, "jar", classifier, jarFile);
+            }
+        }
```
Comment (reinhapa) [inline izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java:205]: I would use an `else if` here instead

File: izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java [hunk 2]
```diff
@@ -224,13 +261,22 @@ private void initMavenProperties(PropertyManager propertyManager)
...
+                    String existingValue = propertyManager.getProperty(propertyName);
+                    if (existingValue != null && existingValue.equals(value))
                     {
-                        getLog().debug("Maven property added: " + propertyName + "=" + value);
+                        getLog().debug("Maven property exists: " + propertyName + "=" + value);
                     }
                     else
                     {
-                        getLog().warn("Maven property " + propertyName + " could not be overridden");
+                        if (propertyManager.addProperty(propertyName, value))
```
Comment (reinhapa) [inline izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java:271]: Here I think using an other `else if` block would be easier to read

**Review by reinhapa — CHANGES_REQUESTED** [review]

**Comment by fransbouwmans** [comment]
It seems that the default behavior of the classifier is not correct yet.
The setDefaultClassifier is testing for "-", which is not the default for a not set classifier. As a result when no classifier is provided the default becomes "installer".
This then is used with the finalName in getJarFile(). In my example the finalName is set to artifact-installer.jar and no classifier set:
in target the installer is created as artifact-installer-installer.jar and is uploaded as artifact-installer.jar.

I suspect that neither finalName and classifier should have a default, such that upon runtime it can be checked whether it is available. I will do some more testing to see if it can be improved.

**Comment by fransbouwmans** [comment]
To get the expected behavior I removed the default value for finalName (i.e. to distinct between an intentional finalName and the default), and then changed the execute to the following initial code:

`	public void execute() throws MojoExecutionException, MojoFailureException {
		if (classifier == null) {
			classifier = "";
		}
		if (project != null) {
			if (project.getPackaging().equals("izpack-jar")) {
				if (!classifier.isEmpty()) {
					getLog().warn("for packaging izpack-jar, classifier is not used");
					classifier = "";
				}
				finalName = project.getBuild().getFinalName();
				getLog().info("packaging izpack-jar, default output: " + finalName);
			} else if (finalName == null) {
				if (classifier.isEmpty()) {
					getLog().warn("default classifier used");
					classifier = "installer";
					enableAttachArtifact = false;
				}
				finalName = project.getBuild().getFinalName() + "-" + classifier;
			} else {
				if (!classifier.isEmpty()) {
					getLog().info("finalName and classifier set, using classifier for finalName");
					finalName = project.getBuild().getFinalName() + "-" + classifier;
				} else {
					enableAttachArtifact = false;
				}
			}
		}
		File jarFile = new File(outputDirectory, finalName + ".jar");
		
		getLog().info("Building installer with name: " + jarFile.getName());
		getLog().info("Installer is " + (enableAttachArtifact ? "attached" : "not attached"));
`

File: izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java (hunk 2 above)
Comment (fransbouwmans) [reply izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java:271]: Note that I tested the plugin by building the demo app.
The unit tests for the plugin currently fail as default behaviour has changed when a finalName and classifier are provided.
I'm not sure how to correct the unit tests (or whether the default behaviour should be corrected to meet the current unit tests).

**Comment by hiteshbosamiya** [comment]
@fransbouwmans, based on the code changes you have done, I think we can make it even simpler for the user. Whether it is `izpack-jar` or `jar` packaging the behavior can be same as below:
1. The default value of the final name will be `${project.build.finalName}`
2. The default value of the classifier will be `"installer"`
3. Based on the above two points, the output directory will have two files - `<finalName>.jar` (which is a module jar) and `<finalName>-<classifier>.jar` (which is an installer jar)
4. The module jar will always be installed into repository
5. The installer jar will get into the repository based on `enableAttachArtifact` flag

**Note**: Based on the above, your plugin `finalName` should be changed to `artifact` as `-installer` will get appended by the plugin.

File: izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java (hunk 2 above)
Comment (hiteshbosamiya) [reply izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java]: Done

File: izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java (hunk 1 above)
Comment (hiteshbosamiya) [reply izpack-maven-plugin/src/main/java/org/izpack/mojo/IzPackNewMojo.java]: Now, the code has changed.

**Comment by fransbouwmans** [comment]
There is still the issue of the finalName provided as a parameter and default classifier, i.e. the parameters are identified as follows:
```
    /**
     * Name of the generated JAR.
     */
    @Parameter( alias = "jarName", property = "jar.finalName", defaultValue = "${project.build.finalName}")
    private String finalName;

    /**
     * Classifier to add to the artifact generated (applicable when packaging
     * type is jar). If given, the artifact is attachable. Furthermore, the
     * output file name gets -<i>classifier</i> as suffix. If this is not given,
     * it will merely be written to the output directory according to the
     * finalName.
     */
    @Parameter( defaultValue = "installer")
    private String classifier;

    /**
     * Whether to attach the generated installer jar to the project artifact if
     * a classifier is specified (applicable when packaging type is jar). This
     * has no effect if no classifier was specified.
     */
    @Parameter( defaultValue = "true")
    private boolean enableAttachArtifact;
```
1. Name of the generated jar is not true. If this value is provided, it is extended with the classifier (which is installer by default). I.e. if the name artifact-installer.jar is provided and no classifier the output file is artifact-installer-installer.jar.
2. Classifier, by the fact that there is a default, the statement "If this is not given it will merely be written to the output directory according to finalName" is never the case as the default is installer.
3. "This has no effect if no classifier was specified" is never true, as there is a default.

**Comment by hiteshbosamiya** [comment]
@fransbouwmans, I agree with your points. We can summarize the logic as below and the code is also changed accordingly:
1. If the packaging type is `izpack-jar` then `classifier` and `enableAttachArtifact` do not have any role to play as we will have only one artifact. If `finalName` is not given then `${project.build.finalName}` is considered as `finalName` and the generated installer file will be "`<${project.build.finalName}>.jar`".
2. If the packaging type is not `izpack-jar` (i.e. it is `jar` or `pom`) and if `finalName` is not provided then the `finalName` will be `${project.build.finalName}` and the installer file will be generated with the name "`<${project.build.finalName}>-<classifier>.jar`" (if `classifier` is not provided then the `classifier` will be set to `installer`). This generated installer file will be attached as an artifact only if `enableAttachArtifact` is set to true.

**Review by reinhapa — APPROVED** [review]

**Comment by hiteshbosamiya** [comment]
@fransbouwmans, good catch, thanks, modified the as suggested.

**Comment by fransbouwmans** [comment]
> @fransbouwmans, good catch, thanks, modified the as suggested.

Verified and no further comments. Thanks

**Review by akuhtz — APPROVED** [review]

**Pull request merged**